In [18]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

In [2]:
data_raw = pd.read_csv("ppid2_Valence_features.csv")
data_raw.shape

/var/folders/f4/6_9y1nb93x72_74wsk9vg8qh0000gn/T/ipykernel_56709/1171576017.py:1: DtypeWarning: Columns (0,1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25,26,27,28,29,30,31,32,33,34,35,36,37,38,39,40,41,42,43,44,45,46,47,48,49,50,51,52,53,54,55,56,57,58,59,60,61,62,63,64,65,66,67,68,69,70,71,72,73,74,75,76,77,78,79,80,81,82,83,84,85,86,87,88,89,90,91,92,93,94,95,96,97,98,99,100,101,102,103,104,105,106,107,108,109,110,111,112,113,114,115,116,117,118,119,120,121,122,123,124,125,126,127,128,129,130,131) have mixed types. Specify dtype option on import or set low_memory=False.
  data_raw = pd.read_csv("ppid2_Valence_features.csv")


(5827, 135)

In [3]:
# removing first row
data = data_raw.iloc[1: , :]

# removing where time has null values
data = data[data['Time'].notnull()]

data.head(5)

Time Time.1      EOG    EOG.1      EOG.2       EOG.3      ACC       ACC.1  \
1    0      4  0.15988   0.2456  521.35161  1643.41561  0.06126  6562.72507   
2    1      5    0.161   0.2456  535.30449  1649.63966  0.06126   6553.4292   
3    2      6  0.17014  0.24593   518.3956  1653.90028  0.06126   6552.9059   
4    3      7  0.17014  0.24593  534.68062  1641.58191  0.06742  6556.01749   
5    4      8  0.17014  0.21327   531.5686  1670.96072  0.08616  6577.66903   

       EDA    EDA.1  ... EEGConn.53 EEGConn.54 EEGConn.55 EEGConn.56  \
1  0.00008  0.00045  ...    0.48763    0.25578    0.01546     0.0168   
2        0        0  ...    0.42995    0.09945    0.04304    0.03403   
3        0        0  ...    0.42875    0.09043    0.05034    0.06566   
4        0        0  ...    0.42099     0.0734    0.05023    0.06085   
5        0  0.00001  ...    0.37806    0.15172    0.40395    0.28963   

  EEGConn.57 EEGConn.58 EEGConn.59 Event Event.1 Event.2  
1    0.13589    0.18419    0.47116   NaN     NaN     NaN  
2    0.05901    0.20058    0.43406   NaN     NaN     NaN  
3     0.0272    0.17715    0.27364   NaN     NaN     NaN  
4    0.02355    0.18303    0.26874   NaN     NaN     NaN  
5    0.02408    0.02495     0.1074   NaN     NaN     NaN  

[5 rows x 135 columns]

In [4]:
column_list = data.columns.values.tolist()
column_list

['Time',
 'Time.1',
 'EOG',
 'EOG.1',
 'EOG.2',
 'EOG.3',
 'ACC',
 'ACC.1',
 'EDA',
 'EDA.1',
 'EDA.2',
 'EDA.3',
 'EDA.4',
 'EMG',
 'EMG.1',
 'EMG.2',
 'EEGth',
 'EEGth.1',
 'EEGth.2',
 'EEGth.3',
 'EEGth.4',
 'EEGth.5',
 'EEGth.6',
 'EEGth.7',
 'EEGth.8',
 'EEGth.9',
 'EEGth.10',
 'EEGth.11',
 'EEGth.12',
 'EEGth.13',
 'EEGa',
 'EEGa.1',
 'EEGa.2',
 'EEGa.3',
 'EEGa.4',
 'EEGa.5',
 'EEGa.6',
 'EEGa.7',
 'EEGa.8',
 'EEGa.9',
 'EEGa.10',
 'EEGa.11',
 'EEGa.12',
 'EEGa.13',
 'EEGb',
 'EEGb.1',
 'EEGb.2',
 'EEGb.3',
 'EEGb.4',
 'EEGb.5',
 'EEGb.6',
 'EEGb.7',
 'EEGb.8',
 'EEGb.9',
 'EEGb.10',
 'EEGb.11',
 'EEGb.12',
 'EEGb.13',
 'EEGg',
 'EEGg.1',
 'EEGg.2',
 'EEGg.3',
 'EEGg.4',
 'EEGg.5',
 'EEGg.6',
 'EEGg.7',
 'EEGg.8',
 'EEGg.9',
 'EEGg.10',
 'EEGg.11',
 'EEGg.12',
 'EEGg.13',
 'EEGConn',
 'EEGConn.1',
 'EEGConn.2',
 'EEGConn.3',
 'EEGConn.4',
 'EEGConn.5',
 'EEGConn.6',
 'EEGConn.7',
 'EEGConn.8',
 'EEGConn.9',
 'EEGConn.10',
 'EEGConn.11',
 'EEGConn.12',
 'EEGConn.13',
 'EEGConn.14

### columns:
1 and 2 time (Time is always in chunks, with features calculated over 4 s windows.), \
3-131 features, \
132-134 triggers. 

Column 1: Time start \
Column 2: Time end \
\
Then the features are in 6 types: \
Column 3-6 EOG (eye movements) \
Column 7-8 ACC (accelerometer, head movement) \
Column 9-13 EDA (skin conductance) \
Column 14-16 EMG (face muscle activity) \
Column 17-72 EEG (eeg, in 4 wavebands, eeg theta, alpha, beta, and gamma) \
Column 73-132 EEG connectivity (in 4 wavebands) \
\
Column 133: Trigger values \
Column 134 : Actual time \
Column 135 : Duration (ignored) \
\
Notes on Column 133: \
Sometimes multiple triggers occur in the same window, in which case they are separated by a colon (:) \
the trigger values are too high by 33024 - i.e. subtract from each trigger 33024 to get the right identity \
Trigger to identity conversion. These triggers identify the START of a task: \
223: Training, negative valence \
224: Training, positive valence \
233: Testing, negative valence \
234: Testing, positive valence \
Trigger 7 identifies the END of a task \

In [5]:
# assigning labels

data['Type'] = 'random'

In [6]:
data['Valence'] = 'random'

In [7]:
# Labeling the data
train_neg = '33247'
train_pos = '33248'
test_neg = '33257'
test_pos = '33258'
end = '33031'

# non null data
non_null_data = data[data['Event'].notnull()]

In [8]:
# cast into string
non_null_data['Event'] = non_null_data['Event'].astype(str)
non_null_data['Event.1'] = non_null_data['Event.1'].astype(str)

/var/folders/f4/6_9y1nb93x72_74wsk9vg8qh0000gn/T/ipykernel_56709/3744656094.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  non_null_data['Event'] = non_null_data['Event'].astype(str)
/var/folders/f4/6_9y1nb93x72_74wsk9vg8qh0000gn/T/ipykernel_56709/3744656094.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  non_null_data['Event.1'] = non_null_data['Event.1'].astype(str)


In [9]:
# iterating all to get start time and end time
train_neg_list = []
train_pos_list = []
test_neg_list = []
test_pos_list = []

tmp_start_time = 0
tmp_end_time = 0
# 0 = not occuring, 1 = train_neg, 2 = train_pos, 3 = test_neg, 4 = test_pos
types = 0

for index, row in non_null_data.iterrows():
    event = row['Event'].split(':')
    time = row['Event.1'].split(':')
    
    if types == 0:
        idx = -1
        # find the start time
        if train_neg in event:
            idx = event.index(train_neg)
            types = 1
        if train_pos in event:
            idx = event.index(train_pos)
            types = 2
        if test_neg in event:
            idx = event.index(test_neg)
            types = 3
        if test_pos in event:
            idx = event.index(test_pos)
            types = 4
        
        if idx != -1:
            tmp_start_time = time[idx]
    else:
        # find ending
        if end in event:
            idx = event.index(end)
            tmp_end_time = time[idx]
            
            duration = (tmp_start_time, tmp_end_time)
            if types == 1:
                train_neg_list.append(duration)
            elif types == 2:
                train_pos_list.append(duration)
            elif types == 3:
                test_neg_list.append(duration)
            else:
                test_pos_list.append(duration)
            
            # reset everything
            tmp_start_time = 0
            tmp_start_time = 0
            types = 0

In [10]:
# distribution

print(len(train_neg_list))
print(len(train_pos_list))
print(len(test_neg_list))
print(len(test_pos_list))

24
24
48
48


In [11]:
# cast into float
data['Time'] = data['Time'].astype(float)
data['Time.1'] = data['Time.1'].astype(float)

In [12]:
# labeling based on (start, end) tuples

# labeling train negatives
for start,end in train_neg_list:
    start = float(start) + 0.5
    end = float(end) + 0.5
    
    data.loc[((data['Time'] > start) & (data['Time.1'] < end)), 'Type'] = 'Training'
    data.loc[((data['Time'] > start) & (data['Time.1'] < end)), 'Valence'] = 'Negative'
    
# labeling train positives
for start,end in train_pos_list:
    start = float(start) + 0.5
    end = float(end) + 0.5
    
    data.loc[((data['Time'] > start) & (data['Time.1'] < end)), 'Type'] = 'Training'
    data.loc[((data['Time'] > start) & (data['Time.1'] < end)), 'Valence'] = 'Positive'
    
# labeling test negatives
for start,end in test_neg_list:
    start = float(start) + 0.5
    end = float(end) + 0.5
    
    data.loc[((data['Time'] > start) & (data['Time.1'] < end)), 'Type'] = 'Testing'
    data.loc[((data['Time'] > start) & (data['Time.1'] < end)), 'Valence'] = 'Negative'
    
# labeling test positives
for start,end in test_pos_list:
    start = float(start) + 0.5
    end = float(end) + 0.5
    
    data.loc[((data['Time'] > start) & (data['Time.1'] < end)), 'Type'] = 'Testing'
    data.loc[((data['Time'] > start) & (data['Time.1'] < end)), 'Valence'] = 'Positive'

In [13]:
# clean data. only the one that have labels

clean_data = data[data['Type'] != 'random']

In [28]:
# casting feature values to float, type and valence to string, removing events

feature_cols = list(clean_data.iloc[:,0:132])# Creates list of all column headers
clean_data[feature_cols] = clean_data[feature_cols].astype(float)

target_cols = ['Type','Valence']
clean_data[target_cols] = clean_data[target_cols].astype(str)

clean_data.drop(columns=['Event', 'Event.1', 'Event.2'], inplace=True)

/var/folders/f4/6_9y1nb93x72_74wsk9vg8qh0000gn/T/ipykernel_56709/1750340855.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  clean_data[feature_cols] = clean_data[feature_cols].astype(float)
/var/folders/f4/6_9y1nb93x72_74wsk9vg8qh0000gn/T/ipykernel_56709/1750340855.py:7: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  clean_data[target_cols] = clean_data[target_cols].astype(str)
/var/folders/f4/6_9y1nb93x72_74wsk9vg8qh0000gn/T/ipykernel_56709/1750340855.py:9: SettingWithCopyWarning: 
A value is trying to 

In [29]:
clean_data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1584 entries, 281 to 5810
Columns: 134 entries, Time to Valence
dtypes: float64(132), object(2)
memory usage: 1.6+ MB


In [30]:
# positive signal

positive = clean_data[clean_data['Valence'] == 'Positive']

In [31]:
positive.head()

Time  Time.1      EOG    EOG.1      EOG.2       EOG.3      ACC  \
281  280.0   284.0  0.03269  0.18852  530.00618  1663.87925  0.07023   
282  281.0   285.0  0.03522  0.16972  527.78995  1650.26793  0.03413   
283  282.0   286.0  0.03522  0.16972  527.84376  1645.88954  0.02725   
284  283.0   287.0  0.03608  0.16972  530.29163  1653.78549  0.02725   
285  284.0   288.0  0.03608  0.15331  526.29944  1646.70320  0.02725   

          ACC.1  EDA    EDA.1  ...  EEGConn.52  EEGConn.53  EEGConn.54  \
281  6531.95043  0.0  0.00003  ...     0.15314     0.03407     0.04494   
282  6524.41338  0.0  0.00001  ...     0.03363     0.02157     0.21674   
283  6522.16183  0.0  0.00004  ...     0.01636     0.01908     0.11374   
284  6520.32162  0.0  0.00003  ...     0.00954     0.00278     0.13529   
285  6518.95560  0.0  0.00002  ...     0.05831     0.03878     0.13753   

     EEGConn.55  EEGConn.56  EEGConn.57  EEGConn.58  EEGConn.59      Type  \
281     0.37336     0.33283     0.19793     0.03080     0.34791  Training   
282     0.26975     0.10365     0.19533     0.11734     0.49382  Training   
283     0.32255     0.11144     0.19649     0.02880     0.43887  Training   
284     0.37243     0.11986     0.09740     0.04926     0.51518  Training   
285     0.46473     0.08722     0.10524     0.01096     0.43206  Training   

      Valence  
281  Positive  
282  Positive  
283  Positive  
284  Positive  
285  Positive  

[5 rows x 134 columns]

In [38]:
# plot positive one time

one_time_start, one_time_end = train_pos_list[0]

data_one_time = clean_data[clean_data['Time'] > (one_time_start + 0.5)]

[9.00200000000018,
 26.99799999999982,
 15.005999999999858,
 15.006000000000085,
 26.998000000000047,
 26.998000000000047,
 9.002000000000407,
 9.001999999999953,
 26.998000000000047,
 26.998000000000047,
 26.997999999999593,
 26.998000000000047,
 9.001999999999953,
 15.005999999999858,
 15.006000000000313,
 15.005999999999858,
 26.996000000000095,
 26.998000000000047,
 15.005999999999858,
 15.006000000000313,
 15.005999999999858,
 26.998000000000047,
 26.998000000000047,
 9.001999999999953,
 9.001999999999953,
 9.002000000000407,
 15.006000000000313,
 15.006000000000313,
 26.996000000000095,
 26.998000000000502,
 15.006000000000313,
 9.001999999999498,
 15.006000000000313,
 9.002000000000407,
 26.996000000000095,
 9.002000000000407,
 9.001999999999498,
 15.006000000000313,
 9.002000000000407,
 15.006000000000313,
 9.002000000000407,
 26.997999999999593,
 9.001999999999498,
 15.006000000000313,
 15.005999999999403,
 9.001999999999498,
 9.002000000000407,
 26.997999999999593]

In [32]:
pos_mean = positive.mean(axis=0)
pos_mean

/var/folders/f4/6_9y1nb93x72_74wsk9vg8qh0000gn/T/ipykernel_56709/1268667612.py:1: FutureWarning: Dropping of nuisance columns in DataFrame reductions (with 'numeric_only=None') is deprecated; in a future version this will raise TypeError.  Select only valid columns before calling the reduction.
  pos_mean = positive.mean(axis=0)


Time          3249.877525
Time.1        3253.877525
EOG              0.028783
EOG.1            0.117969
EOG.2          528.049274
                 ...     
EEGConn.55       0.326850
EEGConn.56       0.173614
EEGConn.57       0.140701
EEGConn.58       0.116952
EEGConn.59       0.389938
Length: 132, dtype: float64

In [19]:
# negative signal

negative = clean_data[clean_data['Valence'] == 'Negative']